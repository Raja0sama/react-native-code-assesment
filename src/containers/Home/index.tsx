import React, {useMemo} from 'react';
import {ScrollView, Text, View} from 'react-native';
import HeaderAvatar from '@components/Header/HeaderAvatar';
import {Card} from '@components/CarCard';
import {useDispatch, useSelector} from 'react-redux';
import {CarState, fetchCars} from '@global/reducers/cars';

const Home = () => {
  const dispatch = useDispatch();
  const {data, loading, error} = useSelector(
    (state: {car: CarState}) => state.car,
  );
  useMemo(() => fetchCars()(dispatch), [dispatch]);

  return (
    <View className="bg-[#f1f4f8] flex-1">
      <HeaderAvatar />
      <ScrollView className="p-4 pb-4">
        <Text className="text-[#A4B0BC] text-xl font-[Nunito-Regular]">
          Explore Tesla Car
        </Text>
        <MainContent data={data} loading={loading} error={error} />
      </ScrollView>
    </View>
  );
};
interface MainContentProps extends CarState {}
const colors = ['#D4D4D4', 'gray', '#23398A', '#AF4A4A'];
const MainContent = (props: MainContentProps) => {
  return (
    <View>
      {props.loading && (
        <View className="p-5">
          <Text>Loading Content ....</Text>
        </View>
      )}
      {props.error && (
        <View className="p-5">
          <Text>error : {props.error}</Text>
        </View>
      )}
      {props.data.map((car, index) => (
        <Card
          testID="main-content"
          key={car.model}
          carDetail={car}
          background={colors[index]}
        />
      ))}

      <View className="h-8" />
    </View>
  );
};

export default Home;
