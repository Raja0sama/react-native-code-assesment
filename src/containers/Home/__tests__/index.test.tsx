import React from 'react';
import Home from '../index';
import {Provider} from 'react-redux';
import store from '@global/index';
import renderer from 'react-test-renderer';
import {render, waitFor} from '@testing-library/react-native';
const ReduxWrapper = ({children}: {children: JSX.Element | [JSX.Element]}) => (
  <Provider store={store}>{children}</Provider>
);
jest.mock('@react-navigation/native');

describe('Home', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ReduxWrapper>
          <Home />
        </ReduxWrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly with fetched Mocked', async () => {
    fetch = jest
      .fn()
      .mockImplementation(() => Promise.resolve({json: () => [response]}));

    const tree = render(
      <ReduxWrapper>
        <Home />
      </ReduxWrapper>,
    );

    await tree.rerender(
      <ReduxWrapper>
        <Home />
      </ReduxWrapper>,
    );

    await waitFor(async () => await expect(tree).toMatchSnapshot());

    tree.unmount();
  });
});

const response = {
  id: 1,
  model: 'Model 3',
  basePrice: '184,000',
  features: [
    {
      id: 1,
      name: 'Performance',
      price: '235,000',
    },
    {
      id: 2,
      name: 'Long Range',
      price: '215,000',
    },
  ],
  exteriorColors: [
    {
      id: 1,
      name: 'Pearl White',
      price: 'Included',
    },
    {
      id: 5,
      name: 'Solid Black',
      price: '4500',
    },
    {
      id: 9,
      name: 'Midnight Silver',
      price: '6000',
    },
    {
      id: 13,
      name: 'Deep Blue',
      price: '6000',
    },
    {
      id: 17,
      name: 'Red Multi-Coat',
      price: '7500',
    },
  ],
  interiorOptions: [
    {
      id: 1,
      name: 'All Black',
      price: 'Included',
    },
    {
      id: 5,
      name: 'Black & White',
      price: '4,500',
    },
  ],
  autopilotOptions: [
    {
      id: 1,
      name: 'Autopilot',
      price: '14,000',
    },
    {
      id: 5,
      name: 'Full self-driving',
      price: '28,000',
    },
  ],
  imageLink:
    'https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PPMR,$W41B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=2&crop=0,0,0,0&version=v0121-078c13b0d202212151058',
};
