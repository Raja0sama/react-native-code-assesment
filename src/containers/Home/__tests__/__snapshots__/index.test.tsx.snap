// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#f1f4f8",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "paddingTop": 8,
        },
        {
          "paddingBottom": 20,
        },
        {
          "display": "flex",
        },
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "center",
        },
        {
          "backgroundColor": "#fff",
        },
        {
          "paddingLeft": 32,
          "paddingRight": 32,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../src/assets/avatar.png",
        }
      }
      style={
        {
          "height": 70,
          "width": 70,
        }
      }
    />
    <Text
      style={
        [
          {
            "fontFamily": "Nunito-ExtraBold",
          },
          {
            "fontSize": 24,
            "lineHeight": 32,
          },
          {
            "paddingLeft": 12,
          },
          {
            "color": "#2E2828",
          },
        ]
      }
    >
      Hello, Sharaf
    </Text>
  </View>
  <RCTScrollView
    style={
      [
        {
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
        {
          "paddingBottom": 16,
        },
      ]
    }
  >
    <View>
      <Text
        style={
          [
            {
              "color": "#A4B0BC",
            },
            {
              "fontSize": 20,
              "lineHeight": 28,
            },
            {
              "fontFamily": "Nunito-Regular",
            },
          ]
        }
      >
        Explore Tesla Car
      </Text>
      <View>
        <View
          style={
            [
              {
                "height": 32,
              },
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Home renders correctly with fetched Mocked 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#f1f4f8",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "paddingTop": 8,
        },
        {
          "paddingBottom": 20,
        },
        {
          "display": "flex",
        },
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "center",
        },
        {
          "backgroundColor": "#fff",
        },
        {
          "paddingLeft": 32,
          "paddingRight": 32,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../src/assets/avatar.png",
        }
      }
      style={
        {
          "height": 70,
          "width": 70,
        }
      }
    />
    <Text
      style={
        [
          {
            "fontFamily": "Nunito-ExtraBold",
          },
          {
            "fontSize": 24,
            "lineHeight": 32,
          },
          {
            "paddingLeft": 12,
          },
          {
            "color": "#2E2828",
          },
        ]
      }
    >
      Hello, Sharaf
    </Text>
  </View>
  <RCTScrollView
    style={
      [
        {
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
        {
          "paddingBottom": 16,
        },
      ]
    }
  >
    <View>
      <Text
        style={
          [
            {
              "color": "#A4B0BC",
            },
            {
              "fontSize": 20,
              "lineHeight": 28,
            },
            {
              "fontFamily": "Nunito-Regular",
            },
          ]
        }
      >
        Explore Tesla Car
      </Text>
      <View>
        <View
          style={
            [
              {
                "paddingBottom": 20,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 20,
              },
            ]
          }
        >
          <Text>
            Loading Content ....
          </Text>
        </View>
        <View
          style={
            [
              {
                "height": 32,
              },
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
